Don't allow partial transactions to commit

cmoel opened the pull request:
This prevents partial transactions from committing.

Before this change, if a consuming application caught an error and didn't rethrow it (or a different error), the transaction would commit, leaving the consuming application to clean up bad data. This moves the `COMMIT` and `ROLLBACK` calls into the `finally` blocks for transactions so consuming applications can implement their own error handling and can't accidentally forget to throw an error if one is caught.

Note: the main change for this PR is 3c983cd. The commits leading to it add a missing development dependency and add prettier.

Closes #92.

ospfranco commented:
Please undo adding `prettier` and `pod-install` and the prettier changes, which apparently ran on xcode internal files as well. Adding prettier should be ok in a different PR. I never add `pod-install` as a dependency and only run it via npx, because it annoys when very frequent updates.

As for the change... I don't quite agree with it, IMHO it is a confusing mechanism to have the transaction commit before the code runs. AFAIK most of other libraries implement the transactions in the same manner. Maybe you can provide some counter examples?

cmoel commented:
@ospfranco I force pushed after removing the `prettier` and `pod-install` changes (and opening #95 to add prettier).

I think understand the hesitation. The transaction hasn't yet been committed to the database while the callback in an application would operate like it did (since it _should_ be committed).

One of the attractive features of react-native-quick-sqlite is that transaction callbacks can return a `QueryResult`, which allows for an insert-and-then-select pattern. I updated my transaction example app with an example: https://github.com/cmoel/react-native-quick-sqlite-transaction-issues/compare/8ab9e10ccc81c8bc9f09d09bebde560ac9a933c1...254f664ee037cd6fa233cf44a3ffc5e50f6733a0.

It's quite possible I'm doing something wrong or want to write my database code in way that isn't a good idea. If that's the case, I appreciate your help and I'm happy to close this PR.

ospfranco commented:
In any case I would go for making things more explicit rather hidden behind option flags. I like the approach of TypeORM:

https://orkhan.gitbook.io/typeorm/docs/transactions

cmoel commented:
Cool, I also generally believe being explicit is better than being implicit since it usually results in easier-to-change code. In many/most cases I'd prefer _any_ error to `ROLLBACK` and any non-error to `COMMIT`. Some of the time, I'd like to catch errors before rolling back. It seemed convenient to write it once and have the library do the right thing instead of having each application handle it.

As an alternative to 8818be5, I considered exposing `commit` and `rollback` functions to the transaction callbacks (in addition to `execute` and `executeAsync`). Is this something you'd be open to a PR for? Would you still like to still support the [built-in `COMMIT` and `ROLLBACK` statements](https://github.com/ospfranco/react-native-quick-sqlite/blob/cea63588d702529c0655408754621cc110275618/src/index.ts#L287-L289), or are you thinking about a change that would always require calling `commit` or `rollback` in application code? I can imagine either implementation and want to do whichever you prefer.

ospfranco commented:
I would add `commit` and `rollback` methods to the tx objects

cmoel commented:
Cool. And to verify, you want to keep the current behavior of rolling back if an error is thrown and committing otherwise?

ospfranco commented:
I think it would be nice to keep it, in case people forget about it, at least there is some prevention mechanism that keeps the data from being partially inserted

cmoel commented:
@ospfranco I force-pushed again. The implementation on this branch exposes `commit` and `rollback` functions to transaction `callback`s. I tried following the current strategy when the database is locked in a transaction by adding a `isFinalized` boolean, setting it and checking it as needed. What do you think about this approach? Anything else that should be changed?

ospfranco commented on src/index.ts:
```diff
@@ -234,21 +239,43 @@ QuickSQLite.transaction = (
         throw e;
       } finally {
         locks[dbName].inProgress = false;
+        locks[dbName].isFinalized = true;
```
where is the isFinalized flag set to false again? Did you write some tests on the sample app?

cmoel replied:
🤦‍♂️ Updated in cc2dfe5.

ospfranco commented:
Please write some tests in the example app, one transaction, two transactions one after the other, failing transaction, etc etc.

cmoel commented:
@ospfranco I updated the example app with some test cases. To get the example up and running, I prefixed `pod-install` with `npx` and cleaned up some of the unused or commented out code. My changes also try to add some structure to the test cases in the app. I'm sure there's a better way to arrange the different test cases but I wanted to avoid any large changes.

ospfranco commented:
beautiful, I will take a more in depth look later and release it. Thanks a lot!

ospfranco commented:
I've modified your code to not have `finalized` as a global variable but rather as a local one. Unfortunately was too dumb with git and somehow this PR is not updating, I will just merge it to main and carry on from there. Thanks for the PR!

ospfranco commented:
published under 6.0.0

cmoel commented:
No worries, thanks a bunch.
